packages/interactor/src/matcher.ts

```diff
@@ -0,0 +1,11 @@
+export interface Matcher<T> {
+  bigtestMatcher: true;
+  match(actual: T): boolean;
+  format(): string;
+}
+
+export type MaybeMatcher<T> = Matcher<T> | T;
+
+export function isMatcher<T>(value: MaybeMatcher<T>): value is Matcher<T> {
+  return value && (value as any).bigtestMatcher === true;
```

Comment:
Is there any way to avoid this explicit `any`?

Reply:
I think I'd prefer just to do a structural match on the methods of the object rather than requiring a `bigTestMatcher` attribute which feels awkward when creating matchers and would probably necessitate a special `createMatcher` function that would do nothing but add the `bigTestMatcher` attribute to the object. In other words, when we get down to defining our own matchers, or especially when users define their own matchers, it's a much simpler story to say:

```ts
export function longerThan(length: number): Matcher<string> {
  return {
    match: (actual: string) => string.length > length,
    format: () => `longer than ${length}`
  };
}
```
instead of 

```ts
import { createMatcher } from '@bigtest/interactor';

export function longerThan(length: number): Matcher<string> {
  return createMatcher({
    match: (actual: string) => string.length > length,
    format: () => `longer than ${length}`
  });
}
```

To answer your original question though, I think that if we just match on the structure of the object and not a boolean attribute, we should be able to just say:

```ts
export function isMatcher<T>(value: unknown): value is Matcher<T> {
  let matcher = value as Matcher<T>
  return matcher && typeof matcher.match === 'function' && typeof matcher.format === 'function';
}
```

Reply:
My reasoning for this was that I wanted to avoid the rare edge case where we accidentally think something is a matcher, but this is probably not ever going to happen, so it doesn't really matter.